File: backend/app/schemas/outcome.py
"""Pydantic validation schemas for Employment Outcome Intelligence."""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.job import EmploymentType
from app.models.outcome import PPITier, RetentionStatus


class TrainingAttributionResponse(BaseModel):
    """Schema for associated completed training course attribution."""

    id: uuid.UUID
    enrollment_id: uuid.UUID
    course_id: uuid.UUID
    course_title: str
    provider_id: uuid.UUID
    provider_name: str
    completed_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PlacementOutcomeCreate(BaseModel):
    """Payload to record a new verified placement outcome."""

    application_id: uuid.UUID = Field(..., description="Unique ID of the hired job application")
    placement_date: date = Field(..., description="Date employment commenced")
    starting_salary_annual: float | None = Field(
        None, ge=0.0, description="Annual starting salary in local currency"
    )
    employment_type: EmploymentType = Field(
        default=EmploymentType.FULL_TIME, description="Type of employment engagement"
    )
    contract_id: uuid.UUID | None = Field(
        None, description="Optional associated Employer Skill Contract ID"
    )


class RetentionUpdatePayload(BaseModel):
    """Payload to update retention milestone and optional feedback."""

    retention_status: RetentionStatus = Field(..., description="Target retention status milestone")
    employer_satisfaction_rating: int | None = Field(
        None, ge=1, le=5, description="Employer satisfaction score from 1 (poor) to 5 (excellent)"
    )
    contract_fulfillment_score: float | None = Field(
        None, ge=0.0, le=100.0, description="Skill contract fulfillment score percentage (0-100)"
    )
    employer_feedback_notes: str | None = Field(
        None, max_length=2000, description="Qualitative feedback from hiring manager"
    )


class EmployerFeedbackPayload(BaseModel):
    """Payload to submit or update employer feedback on a placed graduate."""

    employer_satisfaction_rating: int = Field(
        ..., ge=1, le=5, description="Employer satisfaction score from 1 to 5"
    )
    contract_fulfillment_score: float | None = Field(
        None, ge=0.0, le=100.0, description="Skill contract fulfillment score percentage (0-100)"
    )
    employer_feedback_notes: str | None = Field(
        None, max_length=2000, description="Qualitative feedback notes"
    )


class PlacementOutcomeResponse(BaseModel):
    """Detailed placement outcome response schema."""

    id: uuid.UUID
    application_id: uuid.UUID
    candidate_id: uuid.UUID
    candidate_name: str | None = None
    employer_id: uuid.UUID
    employer_company_name: str | None = None
    job_id: uuid.UUID
    job_title: str | None = None
    contract_id: uuid.UUID | None = None
    contract_title: str | None = None
    placement_date: date
    starting_salary_annual: float | None = None
    employment_type: EmploymentType
    retention_status: RetentionStatus
    contract_fulfillment_score: float | None = None
    employer_satisfaction_rating: int | None = None
    employer_feedback_notes: str | None = None
    verified_by_employer: bool = True
    training_attributions: list[TrainingAttributionResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PlacementOutcomeSummary(BaseModel):
    """Summary representation of a placement record."""

    id: uuid.UUID
    application_id: uuid.UUID
    candidate_id: uuid.UUID
    candidate_name: str | None = None
    employer_company_name: str | None = None
    job_title: str | None = None
    placement_date: date
    employment_type: EmploymentType
    retention_status: RetentionStatus
    starting_salary_annual: float | None = None
    employer_satisfaction_rating: int | None = None
    has_training_attribution: bool = False
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ProviderPerformanceResponse(BaseModel):
    """Institutional PPI breakdown and historical placement metrics for a training provider."""

    provider_id: uuid.UUID
    provider_name: str
    period_start: date | None = None
    period_end: date | None = None
    total_enrolled: int = 0
    total_completed: int = 0
    total_placed: int = 0
    completion_rate: float = 0.0
    placement_rate: float = 0.0
    retention_rate_90d: float = 0.0
    average_starting_salary: float | None = None
    average_employer_rating: float | None = None
    ppi_score: float = 0.0
    ppi_tier: PPITier = PPITier.TIER_3_DEVELOPING
    calculation_breakdown: dict[str, float] = Field(
        default_factory=dict,
        description="Weighted sub-score contributions to the total PPI score",
    )

    model_config = ConfigDict(from_attributes=True)


class ProviderLeaderboardItem(BaseModel):
    """Ranked training provider performance benchmark entry."""

    rank: int
    provider_id: uuid.UUID
    provider_name: str
    ppi_score: float
    ppi_tier: PPITier
    completion_rate: float
    placement_rate: float
    retention_rate_90d: float
    average_starting_salary: float | None = None
    average_employer_rating: float | None = None
    total_graduates: int = 0
    total_placed: int = 0

    model_config = ConfigDict(from_attributes=True)


class SkillPlacementRateInsight(BaseModel):
    """Aggregated placement and wage conversion metrics for a canonical skill."""

    skill_id: uuid.UUID
    skill_name: str
    category: str | None = None
    placement_count: int = 0
    placement_rate: float = 0.0
    average_salary: float | None = None
    retention_rate_90d: float = 0.0


class DistrictOutcomeKPIs(BaseModel):
    """Geographic outcome intelligence metrics for regional planning."""

    district: str
    state: str | None = None
    total_placements: int = 0
    average_salary: float | None = None
    retention_rate_90d: float = 0.0


class OutcomeAnalyticsOverview(BaseModel):
    """Macro-level employment outcome KPIs for government and institutional intelligence."""

    total_placements: int = 0
    placement_rate: float = 0.0
    average_retention_90d: float = 0.0
    average_starting_salary: float | None = None
    average_employer_satisfaction: float | None = None
    total_tracked_candidates: int = 0
    placement_by_employment_type: dict[str, int] = Field(default_factory=dict)
    retention_distribution: dict[str, int] = Field(default_factory=dict)
    district_benchmarks: list[DistrictOutcomeKPIs] = Field(default_factory=list)

File: frontend/src/types/outcome.ts
import { EmploymentType } from "./employer";

export type RetentionStatus =
  | "ACTIVE"
  | "LEFT_WITHIN_30D"
  | "RETAINED_90D"
  | "RETAINED_180D"
  | "TERMINATED";

export type PPITier =
  | "TIER_1_EXCELLENT"
  | "TIER_2_PROFICIENT"
  | "TIER_3_DEVELOPING"
  | "TIER_4_NEEDS_IMPROVEMENT";

export interface TrainingAttribution {
  id: string;
  enrollment_id: string;
  course_id: string;
  course_title: string;
  provider_id: string;
  provider_name: string;
  completed_at: string;
}

export interface PlacementOutcome {
  id: string;
  application_id: string;
  candidate_id: string;
  candidate_name?: string | null;
  employer_id: string;
  employer_company_name?: string | null;
  job_id: string;
  job_title?: string | null;
  contract_id?: string | null;
  contract_title?: string | null;
  placement_date: string;
  starting_salary_annual?: number | null;
  employment_type: EmploymentType;
  retention_status: RetentionStatus;
  contract_fulfillment_score?: number | null;
  employer_satisfaction_rating?: number | null;
  employer_feedback_notes?: string | null;
  verified_by_employer: boolean;
  training_attributions: TrainingAttribution[];
  created_at: string;
  updated_at: string;
}

export interface PlacementOutcomeSummary {
  id: string;
  application_id: string;
  candidate_id: string;
  candidate_name?: string | null;
  employer_company_name?: string | null;
  job_title?: string | null;
  placement_date: string;
  employment_type: EmploymentType;
  retention_status: RetentionStatus;
  starting_salary_annual?: number | null;
  employer_satisfaction_rating?: number | null;
  has_training_attribution: boolean;
  created_at: string;
}

export interface PlacementOutcomeCreate {
  application_id: string;
  placement_date: string;
  starting_salary_annual?: number | null;
  employment_type?: EmploymentType;
  contract_id?: string | null;
}

export interface RetentionUpdatePayload {
  retention_status: RetentionStatus;
  employer_satisfaction_rating?: number | null;
  contract_fulfillment_score?: number | null;
  employer_feedback_notes?: string | null;
}

export interface EmployerFeedbackPayload {
  employer_satisfaction_rating: number;
  contract_fulfillment_score?: number | null;
  employer_feedback_notes?: string | null;
}

export interface ProviderPerformance {
  provider_id: string;
  provider_name: string;
  period_start?: string | null;
  period_end?: string | null;
  total_enrolled: number;
  total_completed: number;
  total_placed: number;
  completion_rate: number;
  placement_rate: number;
  retention_rate_90d: number;
  average_starting_salary?: number | null;
  average_employer_rating?: number | null;
  ppi_score: number;
  ppi_tier: PPITier;
  calculation_breakdown?: Record<string, number>;
}

export interface ProviderLeaderboardItem {
  rank: number;
  provider_id: string;
  provider_name: string;
  ppi_score: number;
  ppi_tier: PPITier;
  completion_rate: number;
  placement_rate: number;
  retention_rate_90d: number;
  average_starting_salary?: number | null;
  average_employer_rating?: number | null;
  total_graduates: number;
  total_placed: number;
}

export interface SkillPlacementRateInsight {
  skill_id: string;
  skill_name: string;
  category?: string | null;
  placement_count: number;
  placement_rate: number;
  average_salary?: number | null;
  retention_rate_90d: number;
}

export interface DistrictOutcomeKPIs {
  district: string;
  state?: string | null;
  total_placements: number;
  average_salary?: number | null;
  retention_rate_90d: number;
}

export interface OutcomeAnalytics {
  total_placements: number;
  placement_rate: number;
  average_retention_90d: number;
  average_starting_salary?: number | null;
  average_employer_satisfaction?: number | null;
  total_tracked_candidates: number;
  placement_by_employment_type: Record<string, number>;
  retention_distribution: Record<string, number>;
  district_benchmarks: DistrictOutcomeKPIs[];
}
